Simulator/Utils/Factory.cpp

```diff
@@ -0,0 +1,120 @@
+#include "RNG/Norm.h"
+
+template <typename T> std::map<std::string, T *(*)(void)> getCreateFunctionForType(T *placeholder)
+{
+   std::map<std::string, T *(*)(void)> createFunctionMap;
+
+   // Register Connections classes
+   if constexpr (std::is_same_v<T, Connections>) {
+      createFunctionMap["ConnStatic"] = &ConnStatic::Create;
+      createFunctionMap["ConnGrowth"] = &ConnGrowth::Create;
+      createFunctionMap["Connections911"] = &Connections911::Create;
+   }
+
+   // Register AllEdges classes
+   else if constexpr (std::is_same_v<T, AllEdges>) {
+      createFunctionMap["AllSpikingSynapses"] = &AllSpikingSynapses::Create;
+      createFunctionMap["AllSTDPSynapses"] = &AllSTDPSynapses::Create;
+      createFunctionMap["AllDSSynapses"] = &AllDSSynapses::Create;
+      createFunctionMap["AllDynamicSTDPSynapses"] = &AllDynamicSTDPSynapses::Create;
+      createFunctionMap["All911Edges"] = &All911Edges::Create;
+   }
+
+   // Register Layout classes
+   else if constexpr (std::is_same_v<T, Layout>) {
+      createFunctionMap["FixedLayout"] = &FixedLayout::Create;
+      createFunctionMap["DynamicLayout"] = &DynamicLayout::Create;
+      createFunctionMap["Layout911"] = &Layout911::Create;
+   }
+
+   // Register AllVertices classes
+   else if constexpr (std::is_same_v<T, AllVertices>) {
+      createFunctionMap["AllLIFNeurons"] = &AllLIFNeurons::Create;
+      createFunctionMap["AllIZHNeurons"] = &AllIZHNeurons::Create;
+      createFunctionMap["All911Vertices"] = &All911Vertices::Create;
+   }
+
+   // Register IRecorder classes
+   else if constexpr (std::is_same_v<T, IRecorder>) {
+      createFunctionMap["XmlRecorder"] = &XmlRecorder::Create;
+      createFunctionMap["XmlGrowthRecorder"] = &XmlGrowthRecorder::Create;
+      createFunctionMap["XmlSTDPRecorder"] = &XmlSTDPRecorder::Create;
+      createFunctionMap["Xml911Recorder"] = &Xml911Recorder::Create;
+#if defined(HDF5)
+      createFunctionMap["Hdf5Recorder"] = &Hdf5Recorder::Create;
+      createFunctionMap["Hdf5GrowthRecorder"] = &Hdf5GrowthRecorder::Create;
+#endif
+   }
+
+   // Register MTRand classes
+   else if constexpr (std::is_same_v<T, MTRand>) {
+      createFunctionMap["MTRand"] = &MTRand::Create;
+      createFunctionMap["Norm"] = &Norm::Create;
+   }
+   return createFunctionMap;
+}
+
+// Template Instantiation of specific types
+template std::map<std::string, Connections *(*)(void)>
+   getCreateFunctionForType<Connections>(Connections *);
+
+template std::map<std::string, AllVertices *(*)(void)>
+   getCreateFunctionForType<AllVertices>(AllVertices *);
+
+template std::map<std::string, IRecorder *(*)(void)>
+   getCreateFunctionForType<IRecorder>(IRecorder *);
+
+template std::map<std::string, AllEdges *(*)(void)> getCreateFunctionForType<AllEdges>(AllEdges *);
+
+template std::map<std::string, Layout *(*)(void)> getCreateFunctionForType<Layout>(Layout *);
+
+template std::map<std::string, MTRand *(*)(void)> getCreateFunctionForType<MTRand>(MTRand *);
```

Comment:
Why are these needed?

Reply:
By placing the declaration in the header and definition in the cpp file I was guarding instantiating the template against any non-declared type. However, in order to make the code more readable I have removed the explicit instantiation and instead added a static_assert to check for unexpected types.